google/cloud/storage/examples/storage_object_samples.cc

```diff
@@ -341,6 +341,71 @@ void ReadEncryptedObject(google::cloud::storage::Client client, int& argc,
   //! [read encrypted object] [END storage_download_encrypted_file]
   (std::move(client), bucket_name, object_name, base64_aes256_key);
 }
+
+void ComposeObject(google::cloud::storage::Client client, int& argc,
+                         char* argv[]) {
+  if (argc < 5) {
+    throw Usage{
+        "compose-object <bucket-name> <destination-object-name>"
+        " <object_1> <object_2> ..."};
+  }
+  auto bucket_name = ConsumeArg(argc, argv);
+  auto destination_object_name = ConsumeArg(argc, argv);
+  std::vector<google::cloud::storage::ComposeSourceObject> compose_objects;
+  while (argc > 1) {
+    auto object_name = ConsumeArg(argc, argv);
+    compose_objects.push_back({object_name});
```

Comment:
nit: I would have used `compose_objects.emplace_back(ConsumeArg(argc, argv));` but both are readable, so meh.

Reply:
Surfacing in case this is something we need to support better.  

In order to make it use `emplace_back()` I need to do the following:
```cpp
auto object_name = google::cloud::storage::ComposeSourceObject{ConsumeArg(argc, argv)};
compose_objects.emplace_back(object_name); // could be inline.
```

This isn't readable, so I decided to keep `push_back()`.


Otherwise, I hit the following compiler error:
2. IIUC, the Issue with the following is it's try passing the Argument and trying to construct the struct using a constructor that it doesn't have.
```cpp
compose_objects.emplace_back(ConsumeArg(argc, argv));
```
```bash
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/ext/new_allocator.h:136:23: error: no matching constructor for initialization of 'google::cloud::storage::v0::ComposeSourceObject'
        { ::new((void *)__p) _Up(std::forward<_Args>(__args)...); }
                             ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/alloc_traits.h:475:8: note: in instantiation of function template specialization '__gnu_cxx::new_allocator<google::cloud::storage::v0::ComposeSourceObject>::construct<google::cloud::storage::v0::ComposeSourceObject, const char *>' requested here
        { __a.construct(__p, std::forward<_Args>(__args)...); }
              ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/7.3.0/../../../../include/c++/7.3.0/bits/vector.tcc:100:21: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<google::cloud::storage::v0::ComposeSourceObject> >::construct<google::cloud::storage::v0::ComposeSourceObject, const char *>' requested here
            _Alloc_traits::construct(this->_M_impl, this->_M_impl._M_finish,
                           ^
/v/google/cloud/storage/examples/storage_object_samples.cc:388:21: note: in instantiation of function template specialization 'std::vector<google::cloud::storage::v0::ComposeSourceObject, std::allocator<google::cloud::storage::v0::ComposeSourceObject> >::emplace_back<const char *>' requested here
    compose_objects.emplace_back(ConsumeArg(argc, argv));
                    ^
/v/google/cloud/storage/object_metadata.h:36:8: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'const char *' to 'const google::cloud::storage::v0::ComposeSourceObject' for 1st argument
struct ComposeSourceObject {
       ^
/v/google/cloud/storage/object_metadata.h:36:8: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'const char *' to 'google::cloud::storage::v0::ComposeSourceObject' for 1st argument
/v/google/cloud/storage/object_metadata.h:36:8: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided
```

Reply:
Oh... I see now.  Leave as-is then.